lib/mixins/acroform.js

```diff
@@ -0,0 +1,317 @@
+        this._endChild(fieldRef);
+      });
+      this._root.data.AcroForm.end();
+    }
+    return this;
+  },
+
+  _endChild (ref) {
+    if (Array.isArray(ref.data.Kids)) {
+      ref.data.Kids.forEach(childRef => {
+        this._endChild(childRef);
+      });
+      ref.end();
+    }
+    return this;
+  },
+
+  field (name, options = {}) {
+    let fieldDict = this._fieldDict(name, options);
+    let fieldRef = this.ref(fieldDict);
+    this._addToParent(fieldRef);
+    return fieldRef;
+  },
+
+  widgetAnnot (name, x, y, w, h, options = {}) {
+    let fieldDict = this._fieldDict(name, options);
+    fieldDict.Subtype = 'Widget';
+    if (fieldDict.F === undefined) {
+      fieldDict.F = 4;    // print the annotation 
+    }
+
+    // Add Field annot to page, and get it's ref
+    this.annotate(x, y, w, h, fieldDict);
+    let annotRef = this.page.annotations[this.page.annotations.length - 1];
+
+    return this._addToParent(annotRef);
+  },
+
+  _addToParent (fieldRef) {
+    let parent = fieldRef.data.Parent;
+    if (parent) {
+      if (!parent.data.Kids) {
+        parent.data.Kids = [];
+      }
+      parent.data.Kids.push(fieldRef);
+    } else {
+      this._root.data.AcroForm.data.Fields.push(fieldRef);
+    }
+    return this;
+  },
+
+  _fieldDict (name, options = {}) {
+    if (!this._acroform) {
+      throw new Error('Call document.initAcroForm() method before adding AcroForm Fields to document');
+    }
+    let opts = Object.assign({}, options);
+    opts = this._resolveFlags(opts);
+    opts = this._resolveJustify(opts);
+    opts = this._resolveFont(opts);
+    opts = this._resolveStrings(opts);
+    opts = this._resolveColors(opts);
+    opts = this._resolveFormat(opts);
+    opts = Object.assign(opts, {
+      T: new String(name),
+    });
+    return opts;
+  },
+
+  _resolveFormat (opts) {
+    if (opts.format && opts.format.type) {
+      let fnKeystroke;
+      let fnFormat;
+      let params;
+      if (FORMAT_SPECIAL[opts.format.type] !== undefined) {
+        fnKeystroke = `AFSpecial_Keystroke`;
+        fnFormat = `AFSpecial_Format`;
+        params = FORMAT_SPECIAL[opts.format.type]
+      } else {
+        let format = opts.format.type.charAt(0).toUpperCase() + opts.format.type.slice(1);
+        fnKeystroke = `AF${format}_Keystroke`;
+        fnFormat = `AF${format}_Format`;
+        params = this._formatParamsAsString(opts.format.params);
+        if (opts.format.type === 'date') {
+          fnKeystroke += 'Ex';
+        }
+      }
+      opts.AA = opts.AA ? opts.AA : {};
+      opts.AA.K = {
+        S: "JavaScript",
+        JS: new String(`${fnKeystroke}(${params});`)
+      };
+      opts.AA.F = {
+        S: "JavaScript",
+        JS: new String(`${fnFormat}(${params});`)
+      };
+    }
+    delete opts.format;
+    return opts;
+  },
+
+  _formatParamsAsString (params) {
+    if (Array.isArray(params)) {
+      return JSON.stringify(params).replace('[', '').replace(']', '')
```

Comment:
Is this safe? What if one of the params contains a `[` or `]` character? Maybe it would be safer to do `params.map(p => JSON.stringify(p)).join(',')`

Reply:
The method has been removed due to changes to support more user friendly format specification.